src/audio/google/google_rtc_audio_processing.c

```diff
@@ -93,6 +108,130 @@ void GoogleRtcFree(void *ptr)
 	return rfree(ptr);
 }
 
+static ALWAYS_INLINE float clamp_rescale(float max_val, float x)
+{
+	float min = -1.0f;
+	float max = 1.0f - 1.0f / max_val;
+
+	return max_val * (x < min ? min : (x > max ? max : x));
+}
+
+static ALWAYS_INLINE float s16_to_float(const char *ptr)
+{
+	float scale = -(float)SHRT_MIN;
+	float x = *(int16_t *)ptr;
+
+	return (1.0f / scale) * x;
+}
+
+static ALWAYS_INLINE void float_to_s16(float x, char *dst)
+{
+	*(int16_t *)dst = (int16_t)clamp_rescale(-(float)SHRT_MIN, x);
+}
+
+static ALWAYS_INLINE float s32_to_float(const char *ptr)
+{
+	float scale = -(float)INT_MIN;
+	float x = *(int32_t *)ptr;
+
+	return (1.0f / scale) * x;
+}
+
+static ALWAYS_INLINE void float_to_s32(float x, char *dst)
+{
+	*(int32_t *)dst = (int16_t)clamp_rescale(-(float)INT_MIN, x);
+}
```

Comment:
I think we already have these conversions in a maths header. @singalsu where can @andyross find existing conversions ?

Reply:
We actually don't, not in the form needed here.  There is a fixed point set of conversion macros, but those are in maximal/64-bit precision for accuracy (they're intended for creating constants) and a pcm_converter library that is a C function that takes as an argument a whole array.  The idea here is to get optimized code that does the fully unrolled de/interleave, format conversion and buffer rollover in one go.  There was some 

Also, to be fair, this is replacing code that also doesn't use the existing converters.  That one was based on xcc/xt-clang intrinsics where this one is C, but both generate the same HiFi/FPU output (actually xt-clang on -O3 will even autovectorize this, though we don't use that mode).

Reply:
@singalsu can we split these out into a common location after merge, it would save a lot of duplication. Also we should plan for SIMD versions of these especially since we are passing in arrays here.